core/drivers/crypto/hisilicon/include/sec_main.h

```diff
@@ -0,0 +1,376 @@
+
+struct bd3_no_scene {
+	uint64_t c_ivin_addr;
+	uint32_t rsvd0;
+	uint32_t rsvd1;
+	uint32_t rsvd2;
+} __packed __aligned(4);
+
+struct bd3_pbkdf2_scene {
+	uint64_t c_ivin_addr;
+
+	/*
+	 * pbkdf2_salt_len: 0~23 bits
+	 * rsvd0: 24~31 bits
+	 */
+	uint32_t pbkdf2_salt_len;
+
+	/*
+	 * c_num: 0~23 bits
+	 * rsvd1: 24~31 bits
+	 */
+	uint32_t c_num;
+
+	/*
+	 * pass_word_len: 0~15 bits
+	 * dk_len: 16~31 bits
+	 */
+	uint32_t pass_word_dk_len;
+} __packed __aligned(4);
+
+struct hisi_sec_bd3_sqe {
+	/*
+	 * type: 0~3 bit
+	 * bd_invalid: 4 bit
+	 * scene: 5~8 bit
+	 * de: 9~10 bit
+	 * src_addr_type: 11~13 bit
+	 * dst_addr_type: 14~16 bit
+	 * mac_addr_type: 17~19 bit
+	 * reserved: 20~31 bits
+	 */
+	uint32_t bd_param;
+
+	/*
+	 * cipher: 0~1 bits
+	 * ci_gen: 2~3 bit
+	 * c_icv_len: 4~9 bit
+	 * c_width: 10~12 bits
+	 * c_key_len: 13~15 bits
+	 */
+	uint16_t c_icv_key;
+
+	/*
+	 * c_mode : 0~3 bits
+	 * c_alg : 4~7 bits
+	 */
+	uint8_t c_mode_alg;
+
+	/*
+	 * nonce_len : 0~3 bits
+	 * huk : 4 bits
+	 * cal_iv_addr_en : 5 bits
+	 * seq : 6 bits
+	 * reserved : 7 bits
+	 */
+	uint8_t huk_iv_seq;
+
+	uint64_t tag;
+	uint64_t data_src_addr;
+	uint64_t a_key_addr;
+	uint64_t a_ivin_addr;
+	uint64_t rsvd;
+	uint64_t c_key_addr;
+
+	/*
+	 * auth: 0~1 bits
+	 * ai_gen: 2~3 bits
+	 * mac_len: 4~8 bits
+	 * akey_len: 9~14 bits
+	 * a_alg: 15~20 bits
+	 * key_sel: 21~24 bits
+	 * ctr_count_mode/sm4_xts: 25~26 bits
+	 * sva_prefetch: 27 bits
+	 * key_wrap_num:28~30 bits
+	 * update_key: 31 bits
+	 */
+	uint32_t auth_mac_key;
+	uint32_t salt;
+	uint16_t auth_src_offset;
+	uint16_t cipher_src_offset;
+
+	/*
+	 * auth_len: 0~23 bit
+	 * auth_key_offset: 24~31 bits
+	 */
+	uint32_t a_len_key;
+
+	/*
+	 * cipher_len: 0~23 bit
+	 * auth_ivin_offset: 24~31 bits
+	 */
+	uint32_t c_len_ivin;
+	uint64_t data_dst_addr;
+	uint64_t mac_addr;
+	union {
+		struct bd3_stream_scene stream_scene;
+		struct bd3_no_scene no_scene;
+		struct bd3_pbkdf2_scene pbkdf2_scene;
+	};
+
+	/*
+	 * done: 0 bit
+	 * icv: 1~3 bit
+	 * csc: 4~6 bit
+	 * flag: 7~10 bit
+	 * reserved: 11~15 bit
+	 */
+	uint16_t done_flag;
+	uint8_t error_type;
+	uint8_t warning_type;
+	uint64_t kek_key_addr;
+	uint32_t counter;
+} __packed __aligned(4);
+
+enum sec_bd_type {
+	BD_TYPE1 = 0x1,
+	BD_TYPE2 = 0x2,
+	BD_TYPE3 = 0x3,
+};
+
+enum sec_cipher_dir {
+	NO_CIPHER,
+	CIPHER_ENCRYPT,
+	CIPHER_DECRYPT,
+	HARDWARE_COPY,
+};
+
+enum sec_auth_dir {
+	NO_AUTH,
+	AUTH_MAC_CALCULATE,
+	AUTH_MAC_VERIFY,
+};
+
+enum sec_ckey_len {
+	CKEY_LEN_128_BIT = 0x0,
+	CKEY_LEN_192_BIT = 0x1,
+	CKEY_LEN_256_BIT = 0x2,
+	CKEY_LEN_SM4 = 0x0,
+	CKEY_LEN_DES = 0x1,
+	CKEY_LEN_3DES_3KEY = 0x1,
+	CKEY_LEN_3DES_2KEY = 0x3,
+};
+
+enum sec_bd_scene {
+	SCENE_NOTHING = 0x0,
+	SCENE_IPSEC = 0x1,
+	SCENE_STREAM = 0x7,
+	SCENE_PBKDF2 = 0x8,
+};
+
+enum sec_bd_seq {
+	DATA_DST_ADDR_DISABLE,
+	DATA_DST_ADDR_ENABLE,
+};
+
+enum sec_buf_type {
+	HISI_FLAT_BUF,
+	HISI_SGL_BUF,
+};
+
+struct hisi_qp *sec_create_qp(uint8_t sq_type);
```

Comment:
nitpicking: if you can provide an inline description comment for this API function (and its arguments), that would be great.